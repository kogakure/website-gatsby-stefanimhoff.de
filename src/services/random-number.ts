export const getRandomNumber = (max: number) =>
  String(Math.floor(Math.random() * Math.floor(max)));
