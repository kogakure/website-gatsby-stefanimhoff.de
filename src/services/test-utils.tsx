import React from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';

import { EmojifyProvider, LocalizedDateProvider } from '../contexts';

const customRender = (
  node: React.ReactElement,
  options?: RenderOptions
): RenderResult => {
  const rendered = render(
    <EmojifyProvider>
      <LocalizedDateProvider>{node}</LocalizedDateProvider>
    </EmojifyProvider>,
    options
  );

  return {
    ...rendered,
    rerender: (ui: React.ReactElement, rerenderOptions?: RenderOptions) =>
      customRender(ui, { container: rendered.container, ...options }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
