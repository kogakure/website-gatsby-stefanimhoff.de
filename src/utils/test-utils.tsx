import React from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { EmojifyProvider, LocalizedDateProvider } from '../contexts';
import { ThemeType } from '../typings/Theme';
import { lightTheme } from '../themes';

interface CustomTheme {
  theme: ThemeType;
}

type RenderOptionsWithTheme = RenderOptions & CustomTheme;

const customRender = (
  node: React.ReactElement,
  options: RenderOptionsWithTheme = { theme: lightTheme }
): RenderResult => {
  const { theme } = options;
  const rendered = render(
    <ThemeProvider theme={theme}>
      <EmojifyProvider>
        <LocalizedDateProvider>{node}</LocalizedDateProvider>
      </EmojifyProvider>
    </ThemeProvider>,
    options
  );

  return {
    ...rendered,
    rerender: (
      ui: React.ReactElement,
      rerenderOptions?: RenderOptionsWithTheme
    ) => customRender(ui, { container: rendered.container, ...options }),
  };
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
