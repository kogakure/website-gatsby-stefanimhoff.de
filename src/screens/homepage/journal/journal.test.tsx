import React from 'react';
import { axe } from 'jest-axe';
import * as Gatsby from 'gatsby';

import { render } from '../../../services/test-utils';

import { Journal } from '.';

const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');

jest.mock('../../../services/random-number', () => ({
  getRandomNumber: () => '3',
}));

useStaticQuery.mockImplementation(() => ({
  homepageJournalPosts: {
    edges: [
      {
        node: {
          excerpt: 'This is an excerpt',
          fields: {
            slug: '/post-1',
          },
          frontmatter: {
            title: 'Post 1',
            date: '2020-05-31T16:00:00.000Z',
            cover: {
              childImageSharp: {
                fluid: {
                  aspectRatio: 0.66,
                  base64: 'data:image/jpeg;base64,XYZ',
                  sizes: '(max-width: 1400px) 100vw, 1400px',
                  src: '/static/cover.jpg',
                  srcSet: '/static/cover.jpg 200w, /static/cover.jpg 400w',
                },
              },
            },
          },
        },
      },
    ],
  },
}));

describe('Homepage', () => {
  describe('Journal', () => {
    test('renders correctly', () => {
      const { container } = render(<Journal />);

      expect(container.firstChild).toMatchSnapshot();
    });

    test('is accessible', async () => {
      const { container } = render(<Journal />);
      const results = await axe(container);

      expect(results).toHaveNoViolations();
    });
  });
});
