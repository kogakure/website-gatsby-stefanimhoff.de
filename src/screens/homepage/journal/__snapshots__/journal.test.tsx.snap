// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Homepage Journal renders correctly 1`] = `
.c1 {
  font-size: clamp(1.50rem,2.9vw,3.25rem);
  font-weight: 900;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 1.0;
  margin-bottom: clamp(1.44rem,3.70vw,1.728rem);
  margin-top: 0;
}

.c1 .autolink {
  background: transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  fill: var(--colorForeground);
  margin-left: clamp( 0.335rem,0.72vw,0.402rem);
  opacity: 0;
  vertical-align: middle;
}

.c1:hover .autolink,
.c1:focus .autolink {
  opacity: 0.5;
}

.c12 {
  font-size: clamp(1.00rem,1.1vw,1.25rem);
  font-weight: 900;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 1.0;
  margin-bottom: clamp( 0.694rem,1.49vw,0.833rem);
  margin-top: 0;
}

.c12 .autolink {
  background: transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  fill: var(--colorForeground);
  margin-left: clamp( 0.335rem,0.72vw,0.402rem);
  opacity: 0;
  vertical-align: middle;
}

.c12:hover .autolink,
.c12:focus .autolink {
  opacity: 0.5;
}

.c4 {
  font-size: clamp(1.00rem,1.1vw,1.25rem);
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  line-height: 1.6;
  margin-bottom: clamp(1.44rem,3.70vw,1.728rem);
  margin-top: 0;
}

.c5 {
  color: var(--colorForeground);
  font-weight: 600;
  -webkit-text-decoration-color: var(--colorForegroundSoft);
  text-decoration-color: var(--colorForegroundSoft);
  -webkit-text-decoration-thickness: 0.1em;
  text-decoration-thickness: 0.1em;
  text-underline-offset: 0.1em;
}

.c5:hover,
.c5:focus {
  -webkit-text-decoration-color: var(--colorAccent);
  text-decoration-color: var(--colorAccent);
}

.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  fill: var(--colorForeground);
  margin-left: 0.2em;
  position: relative;
  top: 0.2em;
  -webkit-transition-duration: 500ms;
  transition-duration: 500ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

a:hover .c6,
a:active .c6,
a:focus .c6 {
  -webkit-transform: translate3D(0.25rem,0,0);
  -ms-transform: translate3D(0.25rem,0,0);
  transform: translate3D(0.25rem,0,0);
}

.c10 {
  cursor: pointer;
}

.c0 {
  grid-column: 2 / -2;
}

.c2 {
  display: grid;
  grid-template-columns: repeat(14,1fr);
}

.c3 {
  grid-column: 1 / 10;
  margin-bottom: clamp(1.44rem,3.70vw,1.728rem);
}

.c7 {
  display: grid;
  gap: clamp( 1.0rem,2.14vw,1.2rem);
  grid-template-columns: repeat(auto-fill,minmax(15rem,1fr));
}

.c8 {
  -webkit-align-self: auto;
  -ms-flex-item-align: auto;
  align-self: auto;
  background-color: var(--colorForegroundFeather);
  border-color: var(--colorForegroundSoft);
  border-style: solid;
  border-width: 1px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin-bottom: 0;
  margin-top: 0;
  padding: clamp( 0.579rem,1.24vw,0.694rem);
  position: relative;
  -webkit-transition-duration: 500ms;
  transition-duration: 500ms;
  -webkit-transition-property: -webkit-transform;
  -webkit-transition-property: transform;
  transition-property: transform;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c8:hover,
.c8:focus {
  -webkit-transform: scale(1.03);
  -ms-transform: scale(1.03);
  transform: scale(1.03);
}

.c8::after {
  box-shadow: 0 0 50px rgba(0,0,0,0.2);
  content: '';
  height: 100%;
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  -webkit-transition-duration: 500ms;
  transition-duration: 500ms;
  -webkit-transition-property: opacity;
  transition-property: opacity;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
  width: 100%;
  z-index: -1;
}

.c8:hover::after,
.c8:focus::after {
  opacity: 1;
}

.c9 {
  color: var(--colorForeground);
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c11 {
  display: block;
  font-size: clamp(0.65rem,0.8vw,0.75rem);
  font-variant: small-caps;
}

@media screen and (prefers-reduced-motion:no-preference) {
  .c1 .autolink {
    -webkit-transition-duration: 500ms;
    transition-duration: 500ms;
    -webkit-transition-property: opacity;
    transition-property: opacity;
    -webkit-transition-timing-function: ease-in-out;
    transition-timing-function: ease-in-out;
  }
}

@media screen and (prefers-reduced-motion:no-preference) {
  .c12 .autolink {
    -webkit-transition-duration: 500ms;
    transition-duration: 500ms;
    -webkit-transition-property: opacity;
    transition-property: opacity;
    -webkit-transition-timing-function: ease-in-out;
    transition-timing-function: ease-in-out;
  }
}

@media (min-width:768px) {
  .c0 {
    grid-column: 3 / -3;
  }
}

@media (min-width:768px) {
  .c8 {
    -webkit-align-self: start;
    -ms-flex-item-align: start;
    align-self: start;
    margin-bottom: 3vw;
    margin-top: 3vw;
  }
}

<section
  class="c0"
>
  <h2
    class="c1"
  >
    Latest Articles
  </h2>
  <header
    class="c2"
  >
    <div
      class="c3"
    >
      <p
        class="c4"
      >
        I am writing about code, design, productivity, self-improvement, films, and books in this Journal and publish a monthly article with recommendations.
      </p>
      <a
        block="0"
        class="c5"
        href="/journal/"
      >
        More
        <span
          class="c6"
        >
          <svg
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M16.172 11l-5.364-5.364 1.414-1.414L20 12l-7.778 7.778-1.414-1.414L16.172 13H4v-2z"
            />
          </svg>
        </span>
      </a>
    </div>
  </header>
  <div
    class="c7"
  >
    <article
      class="c8"
    >
      <a
        aria-label="Post 1"
        class="c9"
        href="/post-1"
      >
        <header>
          <time
            aria-label="Toggle date format"
            class="c10 c11"
            datetime="2020-05-31T16:00:00.000Z"
            role="button"
            tabindex="0"
          >
            May 31st, 2020
          </time>
          <h3
            class="c12"
          >
            Post 1
          </h3>
          This is an excerpt
        </header>
      </a>
    </article>
  </div>
</section>
`;
